README.md
# rest-test

An example REST service.

## Usage

You can start the server locally, assuming you have [leiningen] installed,
with the command `lein ring server`.  This will automatically open a browser
window to `localhost:3000`, although the root URL is not handled so you
will see the message "Not found!"

You can add records using the following curl command:
```
$ curl --data-binary @file.txt http://localhost:3000/records
```

[leiningen]: https://github.com/technomancy/leiningen

## API

### POST /records

POST requests to `/records` should have the "Content-Type" header set to
`text/csv` or `text/plain`.

The POST body should be one line per record, with each field separated by
delimiters.  The delimiters can be `,`, `|`, or ` ` (a space).  Delimiters
may not be doubled.

End-of-line characters should be single newline (`\n`) characters.  A newline
at the end of the file is accepted.

All fields must be present and in the following order (last name can be blank, though):

- **Last Name**
- **First Name**
- **Gender**
- **Favorite Color**
- **Date of Birth** (must be presented in YYYY-MM-DD format)

Fields cannot contain any possible delimiter characters.

### The GET endpoints

There are three GET endpoints:

- `/records/birthdate`
- `/records/gender`
- `/records/name`

All three return the same format response, except for the sort order of the
records.  Here's an example response:

```json
{
  "records": [
    {
      "lastName": "Smith",
      "firstName": "Olive",
      "gender": "enby",
      "favoriteColor": "white",
      "birthdate": "9/2/1896"
    },
    {
      "lastName": "Xmith",
      "firstName": "Olive",
      "gender": "enby",
      "favoriteColor": "white",
      "birthdate": "9/2/1896"
    }
  ]
}
```

The sort order is:

| Endpoint             | Sort Order                                 |
|----------------------|--------------------------------------------|
| `/records/birthdate` | Date of birth, ascending                   |
| `/records/gender`    | Gender ascending, then last name ascending |
| `/records/name`      | Last name, descending                      |

## Development

### TDD

Tests are written test-first fasion using Brian Marick's [Midje].  Midje's
REPL tools are automatically loaded into the REPL's default development
namespace -- you should see a banner explaining how to get its online docs:

```
Run `(doc midje)` for Midje usage.
Run `(doc midje-repl)` for descriptions of Midje repl functions.
```

The most important REPL command is `(autotest)`.  After you run this command,
Midje will watch for file changes and compute which tests need to be re-run,
and run them with nice red or green output.

[suggest-commit] knows how to find new Midje tests and suggest their fact
names for your commit messages.  This'll speed up your inner loop a bit.

[Midje]: https://github.com/marick/Midje
[suggest-commit]: https://github.com/eraserhd/suggest-commit

## License

Copyright © 2017 Jason M. Felice


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-08-04
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-08-04
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/rest-test/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/rest-test/compare/0.1.0...0.1.1


test/rest_test/repl.clj
(ns rest-test.repl
  (:require [clojure.spec.alpha :as s]
            [clojure.spec.gen.alpha :as gen]
            [midje.config]))

;; Convince Midje we are started from the REPL even though we're requiring
;; it from a compiled file -- Midje won't load the online docs otherwise.
(alter-var-root #'midje.config/started-in-repl? (constantly true))
(use 'midje.repl)


test/rest_test/core_test.clj
(ns rest-test.core-test
  (:require [cheshire.core :as json]
            [clojure.spec.alpha :as s]
            [clojure.spec.gen.alpha :as gen]
            [clojure.string :as string]
            [clojure.test.check]
            [clojure.test.check.properties :as prop]
            [midje.sweet :refer :all]
            [ring.mock.request :as mock]
            [rest-test.core :as core]))

(defn- post
  [delimiter & [{:keys [initial-state fields]
                 :or {initial-state #{}
                      fields ["Fabetes" "Joe" "male" "blue" "1997-02-12"]}}]]
  (-> (mock/request :post "/records")
    (mock/header "Content-Type" "text/plain")
    (mock/body (str
                 (string/join delimiter fields) "\n"
                 (string/join delimiter ["Smith" "Jane" "female" "green" "1973-05-06"]) "\n"))
    (assoc :state initial-state)
    core/pure-handler))

(defn holds?
  "Does a property hold?

  A custom midje checker used so that we can see the whole failed test.check
  output when it fails."
  [result]
  (= true (:result result)))

(defmacro property
  "Make a test.check property into a Midje fact."
  ([descr prop]
   `(property ~descr 25 ~prop))
  ([descr trials prop]
   `(fact ~descr
      (clojure.test.check/quick-check ~trials ~prop) => holds?)))

(defn- records
  [endpoint-kind state]
  (let [result (-> (mock/request :get (str "/records/" endpoint-kind))
                 (assoc :state state)
                 core/pure-handler)
        json-result (json/parse-string (:body result) keyword)]
    (:records json-result)))

(defn- in-ascending-order?
  [coll]
  (every? (fn [[a b]] (<= (compare a b) 0)) (partition 2 1 coll)))

(facts "about the REST service"
  (fact "random URLs respond with 404"
    (:status (core/pure-handler (mock/request :get "/somewhere/random"))) => 404)
  (facts "about posts to /records"
    (fact "posts to /records accept comma-separated values"
      (:status (post ",")) => 200
      (:state (post ",")) => #{{:lastName "Fabetes",
                                :firstName "Joe",
                                :gender "male",
                                :favoriteColor "blue",
                                :birthdate "1997-02-12"}
                               {:lastName "Smith",
                                :firstName "Jane",
                                :gender "female"
                                :favoriteColor "green"
                                :birthdate "1973-05-06"}})
    (fact "posts to /records accept pipe-delimited records"
      (:status (post "|")) => 200
      (:state (post "|")) => #{{:lastName "Fabetes",
                                :firstName "Joe",
                                :gender "male",
                                :favoriteColor "blue",
                                :birthdate "1997-02-12"}
                               {:lastName "Smith",
                                :firstName "Jane",
                                :gender "female"
                                :favoriteColor "green"
                                :birthdate "1973-05-06"}})
    (fact "posts to /records accept space-delimited records"
      (:status (post " ")) => 200
      (:state (post " ")) => #{{:lastName "Fabetes",
                                :firstName "Joe",
                                :gender "male",
                                :favoriteColor "blue",
                                :birthdate "1997-02-12"}
                               {:lastName "Smith",
                                :firstName "Jane",
                                :gender "female"
                                :favoriteColor "green"
                                :birthdate "1973-05-06"}})
    (fact "posts to /records preserve existing records"
      (let [initial-state #{{:lastName "Begone"
                             :firstName "Bob"
                             :gender "male"
                             :favoriteColor "mauve"
                             :birthdate "1981-07-01"}}]
        (:state (post "," {:initial-state initial-state})) => #{{:lastName "Fabetes",
                                                                 :firstName "Joe",
                                                                 :gender "male",
                                                                 :favoriteColor "blue",
                                                                 :birthdate "1997-02-12"}
                                                                {:lastName "Smith",
                                                                 :firstName "Jane",
                                                                 :gender "female"
                                                                 :favoriteColor "green"
                                                                 :birthdate "1973-05-06"}
                                                                {:lastName "Begone",
                                                                 :firstName "Bob",
                                                                 :gender "male",
                                                                 :favoriteColor "mauve"
                                                                 :birthdate "1981-07-01"}}))
    (fact "successful posts to /records respond with a JSON success message"
      (:headers (post ",")) => (contains {"Content-Type" #"application/json"})
      (json/parse-string (:body (post ","))) => {"status" "ok"})
    (facts "posts to /records validate input fields"
      (fact "first name must not be empty"
        (:status (post "," {:fields ["Fabetes" "" "male" "blue" "1997-02-12"]})) => 400)
      (fact "last name can be empty"
        (:status (post "," {:fields ["" "Joe" "male" "blue" "1997-02-12"]})) => 200)
      (fact "gender must not be empty"
        (:status (post "," {:fields ["Fabetes" "Joe" "" "blue" "1997-02-12"]})) => 400)
      (fact "favorite color must not be empty"
        (:status (post "," {:fields ["Fabetes" "Joe" "male" "" "1997-02-12"]})) => 400)
      (fact "birthdate must be a valid date"
        (:status (post "," {:fields ["Fabetes" "Joe" "male" "blue" ""]})) => 400
        (:status (post "," {:fields ["Fabetes" "Joe" "male" "blue" "garbage"]})) => 400
        (:status (post "," {:fields ["Fabetes" "Joe" "male" "blue" "1992-02-30"]})) => 400
        (:status (post "," {:fields ["Fabetes" "Joe" "male" "blue" "1992-03-31"]})) => 200)))
  (property "all record retrieval endpoints return all records"
    (prop/for-all [state (s/gen ::core/parsed-body)
                   endpoint-type (gen/elements ["gender" "birthdate" "name"])]
      (= (count state) (count (records endpoint-type state)))))
  (property "all record retrieval endpoints return dates in M/D/YYYY format" ;;
    (prop/for-all [state (s/gen ::core/parsed-body)
                   endpoint-type (gen/elements ["gender" "birthdate" "name"])]
        (every? #(re-matches #"[0-9]{1,2}/[0-9]{1,2}/[0-9]{4}" %)
                (map :birthdate (records endpoint-type state)))))
  (property "/records/gender returns records sorted by gender, then by last name" 50
    (prop/for-all [state (s/gen ::core/parsed-body)]
      (in-ascending-order? (->> (records "gender" state)
                             (map (juxt :gender :lastName))))))
  (property "/records/birthdate returns records sorted ascending by birthdate" 50 ;;
    (prop/for-all [state (s/gen ::core/parsed-body)]
      (in-ascending-order? (->> (records "birthdate" state)
                             (map :birthdate)
                             (map #(.parse (java.text.SimpleDateFormat. "M/d/yyyy") %))))))
  (property "/records/name returns records sorted descending by last name" 50
    (prop/for-all [state (s/gen ::core/parsed-body)]
      (in-ascending-order? (->> (records "name" state)
                             (map :lastName)
                             reverse)))))


test/rest_test/state_test.clj
(ns rest-test.state-test
  (:require [midje.sweet :refer :all]
            [rest-test.state :as state]))

(defn- results
  [requests]
  (let [passed-request-atom (atom nil)
        handler-result-atom (atom nil)
        initial-state (or (:initial-state (first requests)) [])
        handler (state/wrap-state
                  (fn [request]
                    (reset! passed-request-atom request)
                    @handler-result-atom)
                  initial-state)]
    (doall
      (for [{:keys [request handler-result]
             :or {request {},
                  handler-result {}}} requests]
        (do
          (reset! handler-result-atom handler-result)
          (let [result (handler request)]
            {:result result
             :passed-request @passed-request-atom}))))))

(facts "about the `wrap-state` ring middleware"
  (fact "`wrap-state` passes initial state to the wrapped middleware"
    (:passed-request (first (results [{:initial-state 42}]))) => (contains {:state 42}))
  (fact "`wrap-state` returns the wrapped middleware's response"
    (:result (first (results [{:handler-result {:foo "hi, mom!"}}]))) => {:foo "hi, mom!"})
  (fact "`wrap-state` updates the state if returned by the wrapped middleware"
    (:passed-request (last (results [{:initial-state 42,
                                      :handler-result {:state 79}}
                                     {}]))) => (contains {:state 79})))


project.clj
(defproject rest-test "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[cheshire "5.7.1"]
                 [org.clojure/clojure "1.9.0-alpha17"]
                 [org.clojure/spec.alpha "0.1.123"]
                 [ring/ring-core "1.6.2"]
                 [ring/ring-jetty-adapter "1.6.2"]
                 [ring/ring-json "0.4.0"]]
  :profiles {:dev {:dependencies [[midje "1.9.0-alpha6"]
                                  [org.clojure/test.check "0.10.0-alpha2"]
                                  [ring/ring-mock "0.3.1"]]
                   :plugins [[lein-midje "3.2.1"]
                             [lein-ring "0.12.0"]]}}
  :ring {:handler rest-test.core/handler}
  :repl-options {:init-ns rest-test.repl})


src/rest_test/core.clj
(ns rest-test.core
  (:require [clojure.java.io :as io]
            [clojure.spec.alpha :as s]
            [clojure.spec.gen.alpha :as gen]
            [clojure.string :as string]
            [rest-test.state :as state]
            [ring.middleware.json])
  (:import [java.text SimpleDateFormat]))

(def input-files
  ["comma-delimited.txt"
   "pipe-delimited.txt"
   "space-delimited.txt"])

(s/def ::lastName string?)
(s/def ::firstName (s/and string? not-empty))
(s/def ::gender (s/and string? not-empty))
(s/def ::favoriteColor (s/and string? not-empty))

(defn- date-valid?
  [s]
  (let [date-format (doto (SimpleDateFormat. "yyyy-MM-dd")
                      (.setLenient false))]
    (try
      (.parse date-format s)
      true
      (catch Throwable t
        false))))

;; We supply our own generator for ::birthdate because the default one
;; would generate random unicode strings until it finds a valid birthdate.
;; How many, on average, would need to be generated before finding a valid
;; birth date is left as an exercise to the reader.
;;
;; (We still need gen/such-that to prevent things like February 31st.)
(s/def ::birthdate
  (s/spec (s/and string? date-valid?)
          :gen (fn make-birthdate-generator []
                 (gen/such-that
                   date-valid?
                   (gen/fmap
                     (fn [[year month day]]
                       (format "%04d-%02d-%02d" year month day))
                     (gen/tuple
                       (gen/choose 1890 2017)
                       (gen/choose 1 12)
                       (gen/choose 1 31)))))))

(s/def ::record (s/keys :req-un [::firstName
                                 ::gender
                                 ::favoriteColor
                                 ::birthdate]
                        :opt-un [::lastName]))
(s/def ::parsed-body (s/coll-of ::record :kind set?))

(defn- not-found-handler
  [request]
  {:status 404
   :body "Not found!"})

(defn- parse-file
  [body]
  (into #{}
        (comp
          (map #(string/split % #"[,| ]"))
          (map (fn [[lastName firstName gender favoriteColor birthdate]]
                 {:lastName lastName
                  :firstName firstName
                  :gender gender
                  :favoriteColor favoriteColor
                  :birthdate birthdate})))
        (string/split body #"\n")))

(defn- post-handler
  [handler]
  (fn post-handler* [{:keys [request-method uri state body] :as request}]
    (if-not (= [request-method uri] [:post "/records"])
      (handler request)
      (let [parsed-body (parse-file (slurp body))]
        (if (s/valid? ::parsed-body parsed-body)
          {:status 200
           :state (into state parsed-body)
           :body {:status "ok"}}
          {:status 400
           :body {:status "error"
                  :error (s/explain-str ::parsed-body parsed-body)}})))))

(defn- format-date
  [internal-format]
  (let [[year month day] (string/split internal-format #"-")]
    (format "%d/%d/%s" (Long/parseLong month) (Long/parseLong day) year)))

(defn- get-handler
  [handler kind sort-key-fn descending?]
  (fn get-records* [{:keys [request-method uri state] :as request}]
    (if-not (= [request-method uri] [:get (str "/records/" kind)])
      (handler request)
      {:status 200
       :body {:records (cond->> state
                         true        (sort-by sort-key-fn)
                         descending? reverse  ; Didn't actually mean to be cond-descending here,
                                              ; but I couldn't help it.
                         true        (map #(update % :birthdate format-date)))}})))

(def pure-handler
  (-> not-found-handler
    post-handler
    (get-handler "gender" (juxt :gender :lastName) false)
    (get-handler "birthdate" :birthdate false)
    (get-handler "name" :lastName true)
    ring.middleware.json/wrap-json-response))

;; Everything below this line isn't (mechanically) tested

(defn- load-resources
  []
  (let [state (reduce
                (fn [state filename]
                  (into state (parse-file (slurp (io/resource filename)))))
                #{}
                input-files)]
    (when-not (s/valid? ::parsed-body state)
      (s/explain ::parsed-body state))
    state))

(def handler (state/wrap-state pure-handler (load-resources)))
